server/auth/lib.py

```diff
@@ -26,21 +28,47 @@ class UserType(str, enum.Enum):
 
 _SUPERADMIN = "_superadmin"
 _USER = "_user"
+_CREATED_AT = "_created_at"
+_LAST_REQUEST_AT = "_last_request_at"
 
 
-def set_loggedin_user(user_type: UserType, user_key: str):
+def is_session_expired(session):
+    return (
+        datetime.now(timezone.utc)
+        > datetime.fromisoformat(session[_CREATED_AT]) + config.SESSION_LIFETIME
+    ) or (
+        datetime.now(timezone.utc)
+        > datetime.fromisoformat(session[_LAST_REQUEST_AT])
+        + config.SESSION_INACTIVITY_TIMEOUT
+    )
+
+
+def set_loggedin_user(
+    session, user_type: UserType, user_key: str, from_superadmin: bool = False
+):
     session[_USER] = {"type": user_type, "key": user_key}
+    # We don't want to set the created time when a superadmin logs in as
+    # another user, since it was already set when the superadmin logged in.
+    if not from_superadmin:
+        session[_CREATED_AT] = datetime.now(timezone.utc).isoformat()
+    session[_LAST_REQUEST_AT] = datetime.now(timezone.utc).isoformat()
 
 
-def get_loggedin_user() -> Union[Tuple[UserType, str], Tuple[None, None]]:
+def get_loggedin_user(session) -> Union[Tuple[UserType, str], Tuple[None, None]]:
     user = session.get(_USER, None)
-    return (user["type"], user["key"]) if user else (None, None)
+    if not user or is_session_expired(session):
+        return (None, None)
```

Comment:
if session is expired, I would suggest actively expiring the whole session so the cookie gets cleared / renewed.

Reply:
65019224